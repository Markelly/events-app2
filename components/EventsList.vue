<template>
  <div class="events-list">
    <div v-if="isEmptyScreen">No events to show</div>
    <div v-else>
      <SortBy v-if="showAllEvents" data-sort-events @sortEvents="sortEvents" />
      <Card
        v-for="(event, index) in eventsList"
        :key="index"
        :event="event"
        data-card
      />
      <Pagination
        v-if="showPagination"
        :total-pages="totalPages"
        @click="gotoPage"
      />
    </div>
  </div>
</template>

<script>
import { mapActions, mapGetters } from 'vuex'
import { SORT_OPTIONS } from '@/utils/constants'
import Card from '@/components/Card'
import Pagination from '@/components/Pagination'
import SortBy from '@/components/SortBy'

export default {
  name: 'EventsList',
  components: {
    SortBy,
    Pagination,
    Card
  },
  props: {
    showAllEvents: {
      type: Boolean,
      default: false
    }
  },
  data() {
    return {
      eventsList: [],
      page: 0,
      sort: '',
      totalPages: 0,
      sortOptions: SORT_OPTIONS
    }
  },
  computed: {
    ...mapGetters({
      events: 'events',
      pages: 'totalPages'
    }),
    isEmptyScreen() {
      return this.eventsList && !this.eventsList.length
    },
    showPagination() {
      return this.showAllEvents && this.totalPages > 1
    }
  },
  async created() {
    await this.loadList()
  },
  methods: {
    ...mapActions({
      getEvents: 'getEvents',
      getFavorites: 'getFavorites'
    }),
    async loadList(sort) {
      this.showAllEvents
        ? await this.getEvents({ page: this.page, sort })
        : await this.getFavorites()
      this.totalPages = this.pages
      this.eventsList = this.events
    },
    gotoPage(page) {
      this.page = page - 1
      this.loadList()
    },
    sortEvents(sort) {
      this.loadList(sort)
    }
  }
}
</script>

<style scoped lang="scss">
.events-list {
  display: flex;
  flex-direction: column;
  align-items: center;

  & > * {
    width: 100%;
  }
}
</style>
