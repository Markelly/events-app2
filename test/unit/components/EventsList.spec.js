import { createLocalVue, shallowMount } from '@vue/test-utils'
import EventsList from '@/components/EventsList'

const localVue = createLocalVue()
const events = [
  { date: '2020-07-07', title: 'Party' },
  { date: '2020-07-08', title: 'Concert' }
]

describe('<EventsList/>', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(EventsList, {
      localVue,
      propsData: {
        showAllEvents: true
      },
      data: () => {
        return {
          eventsList: events
        }
      }
    })
  })

  test('should load events list', () => {
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.vm.isEmptyScreen).toBeFalsy()
  })

  test('should load empty list', async () => {
    await wrapper.setData({ eventsList: [] })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.vm.isEmptyScreen).toBeTruthy()
  })

  test('should NOT load SortBy components', async () => {
    await wrapper.setProps({ showAllEvents: false })
    expect(wrapper.find('[data-sort-events]').exists()).toBeFalsy()
  })

  test('should show pagination', async () => {
    await wrapper.setData({ totalPages: 3 })
    expect(wrapper.vm.showPagination).toBeTruthy()
  })
})
