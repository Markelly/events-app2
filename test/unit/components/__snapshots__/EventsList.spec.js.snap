// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EventsList/> should NOT load SortBy components 1`] = `
VueWrapper {
  "_emitted": Object {},
  "_emittedByOrder": Array [],
  "isFunctionalComponent": undefined,
}
`;

exports[`<EventsList/> should NOT load SortBy components 2`] = `
VueWrapper {
  "_emitted": Object {},
  "_emittedByOrder": Array [],
  "isFunctionalComponent": undefined,
}
`;

exports[`<EventsList/> should load empty list 1`] = `
VueWrapper {
  "_emitted": Object {},
  "_emittedByOrder": Array [],
  "isFunctionalComponent": undefined,
}
`;

exports[`<EventsList/> should load events list 1`] = `
VueWrapper {
  "_emitted": Object {},
  "_emittedByOrder": Array [],
  "isFunctionalComponent": undefined,
}
`;
